### PR TITLE
Add dependency objects for query services

### LINKED ISSUE #16788: Final pass for dependency objects in public service classes
We may have some public service implementations which should have dependency objects rather than naked parameters, for future-proofing. Specifically, we seem to have at least some cases where a class in core has a dependency object, but not its inheriting class in relational (e.g. QueryContext).

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #16788

### CODE REVIEW

**Comment by ajcvickers** [comment]
Ping @roji Can I get a review of this?

File: src/EFCore.Relational/Query/Internal/RelationalSqlTranslatingExpressionVisitorFactoryDependencies.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    ///         first resolve the object from the dependency injection container, then replace selected
+    ///         services using the 'With...' methods. Do not call the constructor at any point in this process.
+    ///     </para>
+    ///     <para>
+    ///         The service lifetime is <see cref="ServiceLifetime.Singleton"/>. This means a single instance
+    ///         is used by many <see cref="DbContext"/> instances. The implementation must be thread-safe.
+    ///         This service cannot depend on services registered as <see cref="ServiceLifetime.Scoped"/>.
+    ///     </para>
+    /// </summary>
+    public sealed class RelationalSqlTranslatingExpressionVisitorFactoryDependencies
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/Internal/RelationalSqlTranslatingExpressionVisitorFactoryDependencies.cs]: Can we drop `Factory` from the name?

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitorDependencies.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        /// </summary>
+        [EntityFrameworkInternal]
+        public RelationalShapedQueryCompilingExpressionVisitorDependencies(
+            [NotNull] IQuerySqlGeneratorFactory querySqlGeneratorFactory,
+            [NotNull] ISqlExpressionFactory sqlExpressionFactory,
+            [NotNull] IParameterNameGeneratorFactory parameterNameGeneratorFactory)
+        {
+            Check.NotNull(querySqlGeneratorFactory, nameof(querySqlGeneratorFactory));
+            Check.NotNull(querySqlGeneratorFactory, nameof(querySqlGeneratorFactory));
+            Check.NotNull(querySqlGeneratorFactory, nameof(querySqlGeneratorFactory));
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitorDependencies.cs]: Check other argument for null as well

File: src/EFCore.Relational/Query/RelationalShapedQueryOptimizerDependencies.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    ///     <para>
+    ///         The service lifetime is <see cref="ServiceLifetime.Scoped"/>. This means that each
+    ///         <see cref="DbContext"/> instance will use its own instance of this service.
+    ///         The implementation may depend on other services registered with any lifetime.
+    ///         The implementation does not need to be thread-safe.
+    ///     </para>
+    /// </summary>
+    public sealed class RelationalShapedQueryOptimizerDependencies
+    {
+
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalShapedQueryOptimizerDependencies.cs]: Extra line

File: src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs [hunk 4]
```diff
@@ -25,20 +24,22 @@ public class RelationalSqlTranslatingExpressionVisitor : ExpressionVisitor
...
+            Dependencies = dependencies;
+
             _model = model;
             _queryableMethodTranslatingExpressionVisitor = queryableMethodTranslatingExpressionVisitor;
-            _sqlExpressionFactory = sqlExpressionFactory;
-            _memberTranslatorProvider = memberTranslatorProvider;
-            _methodCallTranslatorProvider = methodCallTranslatorProvider;
+            _sqlExpressionFactory = dependencies.SqlExpressionFactory;
+            _memberTranslatorProvider = dependencies.MemberTranslatorProvider;
+            _methodCallTranslatorProvider = dependencies.MethodCallTranslatorProvider;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs]: Don't store the dependencies twice

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.cs [hunk 5]
```diff
@@ -23,20 +23,22 @@ public partial class RelationalShapedQueryCompilingExpressionVisitor : ShapedQue
...
+            : base(queryCompilationContext, dependencies)
         {
-            _querySqlGeneratorFactory = querySqlGeneratorFactory;
-            _sqlExpressionFactory = sqlExpressionFactory;
-            _parameterNameGeneratorFactory = parameterNameGeneratorFactory;
+            RelationalDependencies = relationalDependencies;
+
+            _querySqlGeneratorFactory = relationalDependencies.QuerySqlGeneratorFactory;
+            _sqlExpressionFactory = relationalDependencies.SqlExpressionFactory;
+            _parameterNameGeneratorFactory = relationalDependencies.ParameterNameGeneratorFactory;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.cs]: Don't store the dependencies twice

File: src/EFCore.Relational/Query/SqlExpressionVisitor.cs [hunk 6]
```diff
@@ -3,11 +3,20 @@
...
 using System.Linq.Expressions;
 using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
+using Microsoft.EntityFrameworkCore.Utilities;
 
 namespace Microsoft.EntityFrameworkCore.Query
 {
     public abstract class SqlExpressionVisitor : ExpressionVisitor
     {
+
+        protected SqlExpressionVisitor(QuerySqlGeneratorDependencies dependencies)
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/SqlExpressionVisitor.cs]: If `QuerySqlGeneratorDependencies` is going to be shared between multiple services it should be called `SqlExpressionVisitorDependencies`

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitorDependencies.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitorDependencies.cs]: Query needs a full pass on nullability...but will fix here

File: src/EFCore.Relational/Query/SqlExpressionVisitorDependencies.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        ///     </para>
+        ///     <para>
+        ///         This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///         the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///         any release. You should only use it directly in your code with extreme caution and knowing that
+        ///         doing so can result in application failures when updating to a new Entity Framework Core release.
+        ///     </para>
+        /// </summary>
+        [EntityFrameworkInternal]
+        public SqlExpressionVisitorDependencies(
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressionVisitorDependencies.cs]: This does not seem right. CommandBuilder or SqlGeneratorHelper are not dependencies for SqlExpressionVisitor. SqlExpressionVisitor is not a service either. They are dependencies for QuerySqlGen. QuerySqlGen happens to implement SqlExpressionVisitor. It is same as all other visitors we have which implements ExpressionVisitor. (and we don't have ExpressionVisitorDependency).
SqlExpressionVisitor will remain as base class of double dispatcher which does not require any services given it is abstract class right now. It should be treated as ExpressionVisitor.

Comment (ajcvickers) [reply src/EFCore.Relational/Query/SqlExpressionVisitorDependencies.cs]: Several sub-classes have dependencies. Having this here allows us to add new dependencies, even on the base class, if we need them. I don't think it causes any issues--if no new services are ever used, then we'l never add anything to this.

Comment (ajcvickers) [reply src/EFCore.Relational/Query/SqlExpressionVisitorDependencies.cs]: Resolved by https://github.com/aspnet/EntityFrameworkCore/pull/16847

**Pull request merged**